### PR TITLE
KEYCLOAK-17747 KEYCLOAK-17754 Optimize getClients() calls

### PR DESCRIPTION
**Pull request by hmlnarik**

### CODE REVIEW

**Review by mhajas — COMMENTED** [review]
Looks good to me, however, I am not sure about the DelegateModel in the future. The exception when a model doesn't exist may cause a lot of trouble. Do we plan to remove it in the future? Or maybe replace by some new method `getClientIds(RealmModel)`?

There are some calls to these methods from Migration classes, but these should be alright as there should not be any change in the database during migration right (so no exception should occur)?

However, Should we add the filter for existing clients here (similarly as in ClientsResource):

- https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java#L108
- https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/forms/account/freemarker/model/ApplicationsBean.java#L69
- https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/services/managers/ResourceAdminManager.java#L244
- https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/services/managers/ResourceAdminManager.java#L308

File: server-spi-private/src/main/java/org/keycloak/authorization/store/syncronization/RealmSynchronizer.java [hunk 1]
```diff
@@ -34,17 +35,11 @@ public void synchronize(RealmRemovedEvent event, KeycloakSessionFactory factory)
...
-            ResourceServer resourceServer = storeFactory.getResourceServerStore().findById(clientModel.getId());
-
-            if (resourceServer != null) {
-                String id = resourceServer.getId();
-                //storeFactory.getResourceStore().findByResourceServer(id).forEach(resource -> storeFactory.getResourceStore().delete(resource.getId()));
-                //storeFactory.getScopeStore().findByResourceServer(id).forEach(scope -> storeFactory.getScopeStore().delete(scope.getId()));
-                //storeFactory.getPolicyStore().findByResourceServer(id).forEach(scope -> storeFactory.getPolicyStore().delete(scope.getId()));
-                storeFactory.getResourceServerStore().delete(id);
-            }
+            String id = clientModel.getId();
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply server-spi-private/src/main/java/org/keycloak/authorization/store/syncronization/RealmSynchronizer.java:41]: yes

**Comment by hmlnarik** [comment]
Thanks for the review!

> Looks good to me, however, I am not sure about the DelegateModel in the future. The exception when a model doesn't exist may cause a lot of trouble. Do we plan to remove it in the future? Or maybe replace by some new method `getClientIds(RealmModel)`?

Good point, however it should not be an issue - this should be really an exceptional situation caused by improper transactional level and visible in logs.

> There are some calls to these methods from Migration classes, but these should be alright as there should not be any change in the database during migration right (so no exception should occur)?

> However, Should we add the filter for existing clients here (similarly as in ClientsResource):
> 
> * https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java#L108

Unlikely to occur as above, but should not be preventing export, so yup.

> * https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/forms/account/freemarker/model/ApplicationsBean.java#L69

No, this one has been rejected as it is for account console v1, see [KEYCLOAK-17755](https://issues.redhat.com/browse/KEYCLOAK-17755).

> * https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/services/managers/ResourceAdminManager.java#L244

Yup for similar reason.

> * https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/services/managers/ResourceAdminManager.java#L308

No for similar reason, https://issues.redhat.com/browse/KEYCLOAK-17759

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java [hunk 2]
```diff
@@ -734,9 +736,9 @@ public ClientModel getClientByClientId(RealmModel realm, String clientId) {
         Predicate finalPredicate = builder.and(predicates.toArray(new Predicate[0]));
         queryBuilder.where(finalPredicate).orderBy(builder.asc(root.get("clientId")));
 
-        TypedQuery<ClientEntity> query = em.createQuery(queryBuilder);
+        TypedQuery<String> query = em.createQuery(queryBuilder);
         return closing(paginateQuery(query, firstResult, maxResults).getResultStream())
-                .map(c -> session.clients().getClientById(realm, c.getId()));
+                .map(id -> session.clients().getClientById(realm, id));   // TODO: this discards the obtained client entity and reloads it in getClientById
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java]: done

File: services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java [hunk 3]
```diff
@@ -106,11 +106,10 @@ public static RealmRepresentation exportRealm(KeycloakSession session, RealmMode
 
         if (options.isClientsIncluded()) {
             clients = realm.getClientsStream().collect(Collectors.toList());
-            List<ClientRepresentation> clientReps = new ArrayList<>();
-            for (ClientModel app : clients) {
-                ClientRepresentation clientRep = exportClient(session, app);
-                clientReps.add(clientRep);
-            }
+            List<ClientRepresentation> clientReps = realm.getClientsStream()
```
Comment (mhajas) [inline services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java:109]: We are now calling `realm.getClientsStream()` twice, is this on purpose?

**Review by mhajas — COMMENTED** [review]
Thanks for the update. I added one more comment.

Also, are you sure we don't want to add the filter also to these "deprecated" places? I understand that we don't want to improve performance as it is deprecated, but this means we are adding a new bug/intermittent failure (probably quite rare though), for example, into the old account console. We can leave it as is, I just want to make sure we both understand correctly what it means.

File: services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java (hunk 3 above)
Comment (hmlnarik) [reply services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java:108]: Thanks, my bad :)

**Comment by hmlnarik** [comment]
> Thanks for the update. I added one more comment.
> 
> Also, are you sure we don't want to add the filter also to these "deprecated" places? I understand that we don't want to improve performance as it is deprecated, but this means we are adding a new bug/intermittent failure (probably quite rare though), for example, into the old account console. We can leave it as is, I just want to make sure we both understand correctly what it means.

Yup :-)

**Review by mhajas — APPROVED** [review]
Thank you for the updates @hmlnarik! LGTM

**Pull request merged**